src/server/services/wfs3/qgswfs3handlers.cpp

```diff
@@ -1072,248 +1099,398 @@ void QgsWfs3CollectionsItemsHandler::handleRequest( const QgsServerApiContext &c
+          featureList << feat;
+        i++;
+      }
 
-    // ////////////////////////////////////////////////////////////////////////////////////////////////////
-    // End of input control: inputs are valid, process the request
+      // Count features
+      long matchedFeaturesCount = 0;
+      if ( attrFilters.isEmpty() && filterRect.isNull() )
+      {
+        matchedFeaturesCount = mapLayer->featureCount();
+      }
+      else
+      {
+        if ( filterExpression.isEmpty() )
+        {
+          featureRequest.setNoAttributes();
+        }
 
-    QgsFeatureRequest featureRequest = filteredRequest( mapLayer, context, requestedProperties );
+        featureRequest.setFlags( QgsFeatureRequest::Flag::NoGeometry );
+        featureRequest.setLimit( -1 );
+        features = mapLayer->getFeatures( featureRequest );
 
-    if ( ! filterRect.isNull() )
-    {
-      QgsCoordinateTransform ct( bboxCrs, mapLayer->crs(), context.project()->transformContext() );
-      ct.transform( filterRect );
-      featureRequest.setFilterRect( ct.transform( filterRect ) );
-    }
+        while ( features.nextFeature( feat ) )
+        {
+          matchedFeaturesCount++;
+        }
+      }
 
-    if ( ! attrFilters.isEmpty() )
-    {
+      json data = exporter.exportFeaturesToJsonObject( featureList );
+
+      // Add some metadata
+      data["numberMatched"] = matchedFeaturesCount;
+      data["numberReturned"] = featureList.count();
+      data["links"] = links( context );
+
+      // Current url
+      const QUrl url { context.request()->url() };
 
-      if ( featureRequest.filterExpression() && ! featureRequest.filterExpression()->expression().isEmpty() )
+      // Url without offset and limit
+      QUrl cleanedUrl { url };
+      cleanedUrl.removeQueryItem( QStringLiteral( "limit" ) );
+      cleanedUrl.removeQueryItem( QStringLiteral( "offset" ) );
+
+      QString cleanedUrlAsString { cleanedUrl.toString() };
+
+      if ( ! cleanedUrl.hasQuery() )
       {
-        expressions.push_back( featureRequest.filterExpression()->expression() );
+        cleanedUrlAsString += '?';
       }
-      for ( auto it = attrFilters.constBegin(); it != attrFilters.constEnd(); it++ )
+
+      // Get the self link
+      json selfLink;
+      for ( const auto &l : data["links"] )
       {
-        // Handle star
-        static const QRegularExpression re2( R"raw([^\\]\*)raw" );
-        if ( re2.match( it.value() ).hasMatch() )
+        if ( l["rel"] == "self" )
         {
-          QString val { it.value() };
-          expressions.push_back( QStringLiteral( "\"%1\" LIKE '%2'" ).arg( it.key() ).arg( val.replace( '%', QStringLiteral( "%%" ) ).replace( '*', '%' ) ) );
-        }
-        else
-        {
-          expressions.push_back( QStringLiteral( "\"%1\" = '%2'" ).arg( it.key() ).arg( it.value() ) );
+          selfLink = l;
+          break;
         }
       }
-    }
+      // This should never happen!
+      Q_ASSERT( !selfLink.is_null() );
```

Comment:
I'm not sure that a `Q_ASSERT` is a good idea within the server. What do you think about raising an exception instead?

Reply:
Yes, agreed, even if it will never happen in practice.